File: src/renderer.zig
const std = @import("std");

const sdl = @import("sdl3");

const Color = @import("types.zig").Color;
const Font = @import("types.zig").Font;
const Glyph = @import("types.zig").Glyph;
const GlyphSet = @import("types.zig").GlyphSet;
const Rect = @import("types.zig").Rect;
const TrueType = @import("truetype.zig");
const Window = @import("window.zig").Window;

const Cfg = @import("config.zig").Cfg{};

pub const Renderer = struct {
    window: Window,
    clip: Rect,

    pub fn utf8ToCodepoint(text: []const u8, idx: *usize) u21 {
        var i = idx.*;

        if (i >= text.len) {
            idx.* = text.len;
            return 0xFFFD;
        }

        const b0: u21 = @intCast(text[i]);
        i += 1;

        if (b0 < 0x80) {
            idx.* = i;
            return b0;
        }

        var cp: u21 = 0;

        if (b0 < 0xE0) {
            if (i >= text.len) {
                idx.* = text.len;
                return 0xFFFD;
            }
            cp = ((b0 & 0x1F) << 6) | (@as(u21, @intCast(text[i])) & 0x3F);
            idx.* = i + 1;
            return cp;
        }

        if (b0 < 0xF0) {
            if (i + 1 >= text.len) {
                idx.* = text.len;
                return 0xFFFD;
            }
            cp = ((b0 & 0x0F) << 12) | ((@as(u21, @intCast(text[i])) & 0x3F) << 6) | (@as(u21, @intCast(text[i + 1])) & 0x3F);
            idx.* = i + 2;
            return cp;
        }
        if (i + 2 >= text.len) {
            idx.* = text.len;
            return 0xFFFD;
        }
        cp = ((b0 & 0x07) << 18) | ((@as(u21, @intCast(text[i])) & 0x3F) << 12) | ((@as(u21, @intCast(text[i + 1])) & 0x3F) << 6) | (@as(u21, @intCast(text[i + 2])) & 0x3F);
        idx.* = i + 3;
        return cp;
    }

    pub fn new() !Renderer {
        const win = try Window.new();
        const size = try win.window.getSize();
        const width = size.@"0";
        const height = size.@"1";

        const clip: Rect = .{
            .x = 0,
            .y = 0,
            .width = @intCast(width),
            .height = @intCast(height),
        };

        try win.renderer.setDrawBlendMode(.blend);

        return .{
            .window = win,
            .clip = clip,
        };
    }

    pub fn setClipRect(self: *Renderer, r: Rect) void {
        self.clip = r;
        self.window.renderer.setClipRect(.{
            .h = r.height,
            .w = r.width,
            .x = r.x,
            .y = r.y,
        }) catch {};
    }

    pub fn getSize(self: *Renderer) !struct { usize, usize } {
        return try self.window.window.getSize();
    }

    pub fn loadFont(self: *Renderer, path: []const u8, size: f32) !*Font {
        _ = self;

        const tt_bytes = try std.fs.cwd().readFileAlloc(path, std.heap.page_allocator, .limited(16 * 1024 * 1024));
        errdefer std.heap.page_allocator.free(tt_bytes);

        var tt = try TrueType.load(tt_bytes);

        const vm = tt.verticalMetrics();
        const scale = tt.scaleForPixelHeight(size);
        const height: i32 = @intFromFloat((@as(f32, @floatFromInt(vm.ascent - vm.descent + vm.line_gap))) * scale + 0.5);
        const ascent: i32 = @intFromFloat(@as(f32, @floatFromInt(vm.ascent)) * scale + 0.5); // ADD THIS

        const font = try std.heap.page_allocator.create(Font);
        errdefer std.heap.page_allocator.destroy(font);

        font.* = .{
            .tt_bytes = tt_bytes,
            .tt = tt,
            .size = size,
            .height = height,
            .ascent = ascent, // ADD THIS
            .sets = [_]?*GlyphSet{null} ** 256,
        };

        return font;
    }

    pub fn loadGlyphSet(self: *Renderer, font: *Font, idx: usize) !*GlyphSet {
        var set = try std.heap.page_allocator.create(GlyphSet);
        errdefer std.heap.page_allocator.destroy(set);

        set.width = 128;
        set.height = 128;

        const scale = font.tt.scaleForPixelHeight(font.size + 2);

        while (true) {
            var rgba = try std.heap.page_allocator.alloc(u8, @intCast(set.width * set.height * 4));
            defer std.heap.page_allocator.free(rgba);
            @memset(rgba, 0);

            var pen_x: i32 = 0;
            var pen_y: i32 = font.height;
            var all_fit = true;

            for (0..Cfg.max_glyphset) |i| {
                const cp: u21 = @intCast(idx * Cfg.max_glyphset + i);
                const glyph_index_opt = font.tt.codepointGlyphIndex(cp);
                const glyph_idx: TrueType.GlyphIndex = glyph_index_opt orelse {
                    set.glyphs[i] = Glyph{
                        .w = 0,
                        .h = 0,
                        .off_x = 0,
                        .off_y = 0,
                        .advance = 0,
                        .x0 = 0,
                        .y0 = 0,
                    };
                    continue;
                };

                if (@intFromEnum(glyph_idx) >= font.tt.glyphs_len) {
                    set.glyphs[i] = Glyph{
                        .w = 0,
                        .h = 0,
                        .off_x = 0,
                        .off_y = 0,
                        .advance = 0,
                        .x0 = 0,
                        .y0 = 0,
                    };
                    continue;
                }

                const hm = font.tt.glyphHMetrics(glyph_idx);
                const advance: i32 = @intFromFloat(@as(f32, @floatFromInt(hm.advance_width)) * scale);

                var pixels = std.ArrayListUnmanaged(u8){};
                defer pixels.deinit(std.heap.page_allocator);

                const gbm = font.tt.glyphBitmap(
                    std.heap.page_allocator,
                    &pixels,
                    glyph_idx,
                    scale,
                    scale,
                ) catch |err| {
                    if (err == error.GlyphNotFound) {
                        set.glyphs[i] = Glyph{
                            .w = 0,
                            .h = 0,
                            .off_x = 0,
                            .off_y = 0,
                            .advance = @intCast(advance),
                            .x0 = 0,
                            .y0 = 0,
                        };
                        continue;
                    }
                    return err;
                };

                const w = gbm.width;
                const h = gbm.height;

                if (w == 0 or h == 0) {
                    set.glyphs[i] = Glyph{
                        .w = 0,
                        .h = 0,
                        .off_x = gbm.off_x,
                        .off_y = gbm.off_y,
                        .advance = @intCast(advance),
                        .x0 = 0,
                        .y0 = 0,
                    };
                    continue;
                }

                if (pen_x + w + 2 > set.width) {
                    pen_x = 0;
                    pen_y += font.height + 2;
                }

                if (pen_y + h > set.height) {
                    set.width *= 2;
                    set.height *= 2;
                    all_fit = false;
                    break;
                }

                for (0..h) |yy| {
                    for (0..w) |xx| {
                        const a = pixels.items[yy * w + xx];

                        const dst_x: isize = @as(isize, @intCast(pen_x)) + @as(isize, @intCast(xx));
                        const dst_y: isize = @as(isize, @intCast(pen_y)) + @as(isize, @intCast(yy));

                        if (dst_x < 0 or dst_y < 0) continue;
                        if (dst_x >= set.width or dst_y >= set.height) continue;

                        const ux = @as(usize, @intCast(dst_x));
                        const uy = @as(usize, @intCast(dst_y));

                        const idx_rgba = (uy * @as(usize, @intCast(set.width)) + ux) * 4;
                        rgba[idx_rgba + 0] = 255;
                        rgba[idx_rgba + 1] = 255;
                        rgba[idx_rgba + 2] = 255;
                        rgba[idx_rgba + 3] = a;
                    }
                }

                set.glyphs[i] = Glyph{
                    .w = w,
                    .h = h,
                    .off_x = gbm.off_x,
                    .off_y = gbm.off_y,
                    .advance = @intCast(advance),
                    .x0 = pen_x,
                    .y0 = pen_y,
                };

                pen_x += w + 2;
            }

            if (!all_fit) {
                continue;
            }

            const texture = try std.heap.page_allocator.create(sdl.render.Texture);
            errdefer std.heap.page_allocator.destroy(texture);

            texture.* = try sdl.render.Texture.init(
                self.window.renderer,
                .array_rgba_32,
                .static,
                @intCast(set.width),
                @intCast(set.height),
            );

            set.texture = texture;

            try set.texture.update(null, rgba.ptr, @intCast(set.width * 4));
            try set.texture.setBlendMode(.blend);

            break;
        }

        return set;
    }

    pub fn setFontTabWidth(self: *Renderer, font: *Font, n: i32) !void {
        const state = try self.getGlyphSet(font, '\t');
        state.glyphs['\t'].advance = @intCast(n);
    }

    pub fn getGlyphSet(self: *Renderer, font: *Font, cp: u21) !*GlyphSet {
        const idx = cp >> 8;

        if (idx >= font.sets.len) {
            const replacement_idx: usize = 0xFF;
            if (font.sets[replacement_idx]) |p| return p;
            const set = try self.loadGlyphSet(font, replacement_idx);
            font.sets[replacement_idx] = set;
            return set;
        }

        if (font.sets[idx]) |p| return p;

        const set = try self.loadGlyphSet(font, idx);
        font.sets[idx] = set;
        return set;
    }

    pub fn getFontWidth(self: *Renderer, font: *Font, text: []const u8) !i32 {
        var i: usize = 0;
        var x: i32 = 0;

        while (i < text.len) {
            var cp = utf8ToCodepoint(text, &i);

            if (cp >> 8 >= 256) {
                cp = 0xFFFD;
            }

            const state = try self.getGlyphSet(font, cp);
            const glyph = state.glyphs[cp & 0xFF];
            x += @as(i32, @intCast(glyph.advance));
        }

        return x;
    }

    pub fn getFontHeight(self: *Renderer, font: *Font) i32 {
        _ = self;
        return font.height;
    }

    pub fn getFontTabWidth(self: *Renderer, font: *Font) !i32 {
        const state = try self.getGlyphSet(font, '\t');
        return state.glyphs['\t'].advance;
    }

    pub fn drawText(
        self: *Renderer,
        font: *Font,
        text: []const u8,
        x0: i32,
        y0: i32,
        color: Color,
    ) !i32 {
        var i: usize = 0;
        var x = x0;
        const y = y0;

        while (i < text.len) {
            var cp = utf8ToCodepoint(text, &i);

            if (cp == 0xFFFD and i >= text.len) break;

            if (cp >> 8 >= 256) {
                cp = 0xFFFD;
            }

            const state = try self.getGlyphSet(font, cp);
            const glyph = state.glyphs[cp & 0xFF];

            if (glyph.w > 0 and glyph.h > 0) {
                const src = sdl.rect.FRect{
                    .x = @floatFromInt(glyph.x0),
                    .y = @floatFromInt(glyph.y0),
                    .w = @floatFromInt(glyph.w),
                    .h = @floatFromInt(glyph.h),
                };

                const dst = sdl.rect.FRect{
                    .x = @floatFromInt(x + glyph.off_x),
                    .y = @floatFromInt(y + font.ascent + glyph.off_y),
                    .w = @floatFromInt(glyph.w),
                    .h = @floatFromInt(glyph.h),
                };

                try state.texture.setColorMod(color.r, color.g, color.b);
                try state.texture.setAlphaMod(color.a);

                try self.window.renderer.renderTexture(state.texture.*, src, dst);
            }

            x += @as(i32, @intCast(glyph.advance));
        }
        return x;
    }

    pub fn freeFont(_: *Renderer, font: *Font) void {
        for (font.sets) |set| {
            if (set) |p| {
                p.texture.deinit();
                std.heap.page_allocator.destroy(p.texture);
                std.heap.page_allocator.destroy(p);
            }
        }
        std.heap.page_allocator.free(font.tt_bytes);
        std.heap.page_allocator.destroy(font);
    }

    pub fn drawImage(
        self: *Renderer,
        tex: *sdl.render.Texture,
        sub: Rect,
        x: i32,
        y: i32,
        color: Color,
    ) !void {
        try tex.setColorMod(color.r, color.g, color.b);
        try tex.setAlphaMod(color.a);

        const src = sdl.rect.FRect{
            .x = @floatFromInt(sub.x),
            .y = @floatFromInt(sub.y),
            .w = @floatFromInt(sub.width),
            .h = @floatFromInt(sub.height),
        };

        const dest = sdl.rect.FRect{
            .x = @floatFromInt(x),
            .y = @floatFromInt(y),
            .w = @floatFromInt(sub.width),
            .h = @floatFromInt(sub.height),
        };

        try self.window.renderer.renderTexture(tex.*, src, dest);
    }

    pub fn drawRect(self: *Renderer, rect: Rect, c: Color) void {
        self.window.renderer.setDrawColor(.{ .a = c.a, .r = c.r, .g = c.g, .b = c.b }) catch {
            std.debug.print("Error setting draw color in fill rect", .{});
            return;
        };
        const r = sdl.rect.FRect{
            .x = @floatFromInt(rect.x),
            .y = @floatFromInt(rect.y),
            .w = @floatFromInt(rect.width),
            .h = @floatFromInt(rect.height),
        };

        self.window.renderer.renderFillRect(r) catch {
            std.debug.print("Error drawing fill rect", .{});
            return;
        };
    }
};

File: src/renderer.c
#include <stdio.h>
#include <stdbool.h>
#include <assert.h>
#include <math.h>
#include "lib/stb/stb_truetype.h"
#include "renderer.h"

#define MAX_GLYPHSET 256

typedef struct
{
  SDL_Texture *texture; // GPU texture for atlas
  stbtt_bakedchar glyphs[256];
  int width, height;
} GlyphSet;

struct RenFont
{
  void *data;
  stbtt_fontinfo stbfont;
  GlyphSet *sets[MAX_GLYPHSET];
  float size;
  int height;
};

SDL_Window *window;
SDL_Renderer *renderer;
static RenRect clip;

static void *check_alloc(void *ptr)
{
  if (!ptr)
  {
    fprintf(stderr, "Fatal error: memory allocation failed\n");
    exit(-1);
  }
  return ptr;
}

static const char *utf8_to_codepoint(const char *p, unsigned *dst)
{
  unsigned res, n;
  switch (*p & 0xf0)
  {
  case 0xf0:
    res = *p & 0x07;
    n = 3;
    break;
  case 0xe0:
    res = *p & 0x0f;
    n = 2;
    break;
  case 0xd0:
  case 0xc0:
    res = *p & 0x1f;
    n = 1;
    break;
  default:
    res = *p;
    n = 0;
    break;
  }
  while (n--)
  {
    res = (res << 6) | (*(++p) & 0x3f);
  }
  *dst = res;
  return p + 1;
}

void ren_init(SDL_Window *win, SDL_Renderer *ren)
{
  assert(win);
  assert(ren);
  window = win;
  renderer = ren;

  int w, h;
  SDL_GetWindowSize(win, &w, &h);
  clip = (RenRect){0, 0, w, h};

  SDL_SetRenderDrawBlendMode(renderer, SDL_BLENDMODE_BLEND);
}

void ren_update_rects(RenRect *rects, int count)
{
  SDL_RenderPresent(renderer);
  // SDL_UpdateWindowSurfaceRects(window, (SDL_Rect *)rects, count);
  static bool initial_frame = true;
  if (initial_frame)
  {
    SDL_ShowWindow(window);
    initial_frame = false;
  }
}

void ren_set_clip_rect(RenRect r)
{
  clip = r;
  SDL_SetRenderClipRect(renderer, &(SDL_Rect){
                                      (int)r.x,
                                      (int)r.y,
                                      (int)r.width,
                                      (int)r.height,
                                  });
}

void ren_get_size(int *w, int *h)
{
  SDL_GetWindowSize(window, w, h);
}

static GlyphSet *load_glyphset(RenFont *font, int idx)
{
  GlyphSet *set = check_alloc(calloc(1, sizeof(GlyphSet)));

  int width = 128;
  int height = 128;

retry:
  uint8_t *alpha = malloc(width * height);
  if (!alpha)
    return NULL;

  float s =
      stbtt_ScaleForMappingEmToPixels(&font->stbfont, 1) /
      stbtt_ScaleForPixelHeight(&font->stbfont, 1);

  int res = stbtt_BakeFontBitmap(
      font->data,
      0,
      font->size * s,
      alpha,
      width,
      height,
      idx * 256,
      256,
      set->glyphs);

  if (res < 0)
  {
    free(alpha);
    width *= 2;
    height *= 2;
    goto retry;
  }

  set->width = width;
  set->height = height;

  // Expand to RGBA8888 (white text with alpha)
  uint8_t *rgba = malloc(width * height * 4);
  if (!rgba)
  {
    free(alpha);
    free(set);
    return NULL;
  }

  for (int i = 0; i < width * height; i++)
  {
    rgba[i * 4 + 0] = 255;      // R
    rgba[i * 4 + 1] = 255;      // G
    rgba[i * 4 + 2] = 255;      // B
    rgba[i * 4 + 3] = alpha[i]; // A
  }

  free(alpha);

  // Create SDL3 texture
  set->texture = set->texture = SDL_CreateTexture(renderer,
    SDL_PIXELFORMAT_ABGR8888,
    SDL_TEXTUREACCESS_STATIC, width, height);

  if (!set->texture)
  {
    fprintf(stderr, "SDL_CreateTexture failed: %s\n", SDL_GetError());
    free(rgba);
    free(set);
    return NULL;
  }

  SDL_UpdateTexture(set->texture, NULL, rgba, width * 4);
  SDL_SetTextureBlendMode(set->texture, SDL_BLENDMODE_BLEND);

  free(rgba);

  // Fix metrics: baseline alignment
  int ascent, descent, linegap;
  stbtt_GetFontVMetrics(&font->stbfont, &ascent, &descent, &linegap);
  float scale = stbtt_ScaleForMappingEmToPixels(&font->stbfont, font->size);
  int baseline = ascent * scale + 0.5;

  for (int i = 0; i < 256; i++)
  {
    set->glyphs[i].yoff += baseline;
    set->glyphs[i].xadvance = floor(set->glyphs[i].xadvance);
  }

  return set;
}

static GlyphSet *get_glyphset(RenFont *font, int codepoint)
{
  int idx = (codepoint >> 8) % MAX_GLYPHSET;
  if (!font->sets[idx])
  {
    font->sets[idx] = load_glyphset(font, idx);
  }
  return font->sets[idx];
}

RenFont *ren_load_font(const char *filename, float size)
{
  RenFont *font = check_alloc(calloc(1, sizeof(RenFont)));
  font->size = size;

  FILE *fp = fopen(filename, "rb");
  if (!fp)
    return NULL;

  fseek(fp, 0, SEEK_END);
  int buf_size = ftell(fp);
  fseek(fp, 0, SEEK_SET);

  font->data = check_alloc(malloc(buf_size));
  fread(font->data, 1, buf_size, fp);
  fclose(fp);

  if (!stbtt_InitFont(&font->stbfont, font->data, 0))
  {
    free(font->data);
    free(font);
    return NULL;
  }

  int ascent, descent, linegap;
  stbtt_GetFontVMetrics(&font->stbfont, &ascent, &descent, &linegap);
  float scale = stbtt_ScaleForMappingEmToPixels(&font->stbfont, size);

  font->height = (ascent - descent + linegap) * scale + 0.5;

  // Force loading first glyph set so \t and \n exist
GlyphSet *gset = get_glyphset(font, 0);

if (gset) {
    stbtt_bakedchar *g = gset->glyphs;
    g['\t'].x1 = g['\t'].x0;
    g['\n'].x1 = g['\n'].x0;
}


  return font;
}

void ren_free_font(RenFont *font)
{
  for (int i = 0; i < MAX_GLYPHSET; i++)
  {
    GlyphSet *set = font->sets[i];
    if (set)
    {
      if (set->texture)
        SDL_DestroyTexture(set->texture);
      free(set);
    }
  }
  free(font->data);
  free(font);
}

void ren_set_font_tab_width(RenFont *font, int n)
{
  GlyphSet *set = get_glyphset(font, '\t');
  set->glyphs['\t'].xadvance = n;
}

int ren_get_font_tab_width(RenFont *font)
{
  GlyphSet *set = get_glyphset(font, '\t');
  return set->glyphs['\t'].xadvance;
}

int ren_get_font_width(RenFont *font, const char *text)
{
  int x = 0;
  const char *p = text;
  unsigned codepoint;
  while (*p)
  {
    p = utf8_to_codepoint(p, &codepoint);
    GlyphSet *set = get_glyphset(font, codepoint);
    stbtt_bakedchar *g = &set->glyphs[codepoint & 0xff];
    x += g->xadvance;
  }
  return x;
}

int ren_get_font_height(RenFont *font)
{
  return font->height;
}

static inline RenColor blend_pixel(RenColor dst, RenColor src)
{
  int ia = 0xff - src.a;
  dst.r = ((src.r * src.a) + (dst.r * ia)) >> 8;
  dst.g = ((src.g * src.a) + (dst.g * ia)) >> 8;
  dst.b = ((src.b * src.a) + (dst.b * ia)) >> 8;
  return dst;
}

static inline RenColor blend_pixel2(RenColor dst, RenColor src, RenColor color)
{
  src.a = (src.a * color.a) >> 8;
  int ia = 0xff - src.a;
  dst.r = ((src.r * color.r * src.a) >> 16) + ((dst.r * ia) >> 8);
  dst.g = ((src.g * color.g * src.a) >> 16) + ((dst.g * ia) >> 8);
  dst.b = ((src.b * color.b * src.a) >> 16) + ((dst.b * ia) >> 8);
  return dst;
}

#define rect_draw_loop(expr)      \
  for (int j = y1; j < y2; j++)   \
  {                               \
    for (int i = x1; i < x2; i++) \
    {                             \
      *d = expr;                  \
      d++;                        \
    }                             \
    d += dr;                      \
  }

void ren_draw_rect(RenRect rect, RenColor c)
{
  SDL_SetRenderDrawColor(renderer, c.r, c.g, c.b, c.a);
  SDL_FRect frect = {
      (float)rect.x, (float)rect.y, (float)rect.width, (float)rect.height};
  SDL_RenderFillRect(renderer, &frect);
}

void ren_draw_image(SDL_Texture *tex, RenRect sub, int x, int y, RenColor color)
{
  SDL_SetTextureColorMod(tex, color.r, color.g, color.b);
  SDL_SetTextureAlphaMod(tex, color.a);

  SDL_FRect src = {(float)sub.x, (float)sub.y, (float)sub.width, (float)sub.height};
  SDL_FRect dst = {(float)x, (float)y, (float)sub.width, (float)sub.height};

  SDL_RenderTexture(renderer, tex, &src, &dst);
}

int ren_draw_text(RenFont *font, const char *text, int x, int y, RenColor color)
{
  const char *p = text;
  unsigned codepoint;

  while (*p)
  {
    p = utf8_to_codepoint(p, &codepoint);

    GlyphSet *set = get_glyphset(font, codepoint);
    stbtt_bakedchar *g = &set->glyphs[codepoint & 0xFF];

    int w = g->x1 - g->x0;
    int h = g->y1 - g->y0;

    if (w > 0 && h > 0)
    {
      SDL_FRect src = {
          (float)g->x0,
          (float)g->y0,
          (float)w,
          (float)h};

      SDL_FRect dst = {
          (float)(x + g->xoff),
          (float)(y + g->yoff),
          (float)w,
          (float)h};

      SDL_SetTextureColorMod(set->texture, color.r, color.g, color.b);
      SDL_SetTextureAlphaMod(set->texture, color.a);

      SDL_RenderTexture(renderer, set->texture, &src, &dst);
    }

    x += g->xadvance;
  }

  return x;
}
